graylog2-web-interface/src/views/components/sidebar/fields/FieldsOverview.jsx

```diff
@@ -0,0 +1,125 @@
+// @flow strict
+import * as React from 'react';
+import { useReducer } from 'react';
+import PropTypes from 'prop-types';
+import { List as ImmutableList } from 'immutable';
+import styled, { type StyledComponent } from 'styled-components';
+
+import type { ViewMetaData as ViewMetadata } from 'views/stores/ViewMetadataStore';
+import FieldTypeMapping from 'views/logic/fieldtypes/FieldTypeMapping';
+import { type ThemeInterface } from 'theme';
+
+import { Button } from 'components/graylog';
+import List from './List';
+import ChangeMode from './ChangeMode';
+
+type Props = {
+  viewMetadata: ViewMetadata,
+  listHeight: number,
+  fields: ImmutableList<FieldTypeMapping>,
+  allFields: ImmutableList<FieldTypeMapping>,
+};
+
+const Container: StyledComponent<{}, ThemeInterface, HTMLDivElement> = styled.div`
+  white-space: break-spaces;
+`;
+
+const FilterForm = styled.form`
+  display: flex;
+`;
+
+const FilterInputWrapper = styled.div`
+  flex-grow: 1;
+  margin-right: 5px;
+`;
+
+const FilterInput = styled.input`
+  width: 100%;
+`;
+
+const FieldListModes = styled.div`
+  margin-top: 5px;
+  margin-bottom: 0;
+`;
+
+const reducer = (state, action: {type: string, payload?: { mode?: string, filter?: string }}): { filter: ?string, currentMode: string } => {
+  const { payload = {} } = action;
+  const { mode, filter } = payload;
+  switch (action.type) {
+    case 'search':
+      return { ...state, filter: filter };
+    case 'searchReset':
+      return { ...state, filter: undefined };
+    case 'changeMode':
+      return mode ? { ...state, currentMode: mode } : state;
+    default:
+      return state;
+  }
+};
+
+const FieldsOverview = ({ allFields, fields, viewMetadata, listHeight }: Props) => {
+  const [{ filter, currentMode }, dispatch] = useReducer(reducer, { filter: undefined, currentMode: 'current' });
```

Comment:
Using a reducer here overcomplicates things, as the field category and the filter are not related. They could be handled by using two separate, independent states.